### ISSUE TITLE
file system corruption in the SD CARD

### ISSUE THREAD

**Issue by RAWJUNGLE** [democloid/picoTracker#7]
when trying to re-upload the samples(new samples), the file system crashes and the SD CARD no longer works.

**Comment by democloid** [comment]
What do you mean by "re-download"? can you describe in detail the steps you're doing?

**Comment by RAWJUNGLE** [comment]
1. Upload samples to SD-Card
2. Upload to picotraker
3. Used samples
4. Off pikotracker
5. Upload new samples 
6. Corrupt file system 

When trying to format the SD card, I get the following errors![image](https://github.com/democloid/picoTracker/assets/108550345/a1287937-1060-46ca-84a1-2c1b4ba5d89a)

**Comment by democloid** [comment]
Is it possible you have a bad SD card? the SD card is not mounted as you would in an OS, but read or written to with each operation as needed.
The only write operations that happen are when saving a project, so it's very unlikely that the mere act of loading samples may corrupt the SD card.

**Comment by RAWJUNGLE** [comment]
3 new SD-Card Sun Disk 16gb corrupted on file system. 

I formatted on terminal, But the problem won't go away.

Two of my friends who have collected have the same problem.

**Comment by maks** [comment]
@RAWJUNGLE I just saw your other issue, so I thought I'd better first check which version of the picoTracker you are using? The "desktop" build or the "portable" using the custom PCB?

**Comment by RAWJUNGLE** [comment]
The problem is at the code level, it doesn't depend on the hardware in any way. I built on a breadboard and got these problems. Both with file count and file system problem. I checked this on 3 sd cards and several of my friends checked on their schematics, both breadboard and PCB. They have the same thing.

When you first load the samples, everything works fine, but when you want to load again, you get a file system problem on the sd card.

**Comment by maks** [comment]
@RAWJUNGLE I did some quick tests with my desktop build picotracker which uses a standard RPI pico board using a brand new San Disk Ultra 32GB sdcard to see if I could replicate the issues you reported.


First off I was able to copy across large numbers of sample files, you can see the listing here:
[filelist.txt](https://github.com/democloid/picoTracker/files/13161092/filelist.txt) 
onto the sdcard in the `samplelib` folder as well as a few samples from the lgpt "10k" demo

and then do the following steps:
*  create a new picotracker project on device
* open the folders with the samples
* listen to one of the 10k samples (bassdrum.wav) and import it into the project
* save the project
* turn off picotracker
* put the sdcard back into usb sdcard reader attached to my laptop (running ubuntu 22.04) and successfully read the sdcard on the laptop
* copy more samples from laptop onto the sdcard
* put the card back into picotracker, listen to and then import another sample into the project
* put the card back into laptop reader and browse and read files on sdcard with no issues

So I can't replicate any issue with sdcard filesystem corruption with the above tests.

BUT I did come across several bugs:
* listening and then importing even a tiny sample causes all further playback in the project to be corrupted
* trying to list a directory with over a hundred files in it on the picotracker first takes a very long time 10-15 just to go into the directory listing after pressing `A` to go into it. And then the listing is truncated for me after about 30 entries (sorry I was in hurry and didn't note down exactly how many) with a `F` char being displayed on the last 2 lines  instead of the actual file names.
* the second time around importing a sample actually killed all audio playback. Picotracker still went into playback mode (ie going into the project page to click `save song` would give an error message of `not during playback` but no sound was produced and the playback marker `>` were not being drawn on the song or phrase columns.

All of the above were fixed with a restart. At least 2 of the above are already tracked in #6 so I'll update there with my findings.

So tl;dr there definitely are issues with browsing directories with large numbers of sample files as well as issues importing samples, but I couldn't so far see any evidence of filesystem corruption.

@democloid One thing to note though, is that on importing samples lgpt and hence picotracker does also write the sample file into the projects `samples` sub directory, so I guess at least in theory that could somehow corrupt the filesystem if invalid data was first read during the sample import process, but thats only a guess as I can't reproduce the symptoms in the original report here.

**Comment by RAWJUNGLE** [comment]
Format your SD card, Fat32 format, upload this folder via a link to the root of your SD card. And try working with LGPT, then disable LGPT, and add more sounds to the samplelib directory. 

https://dropmefiles.com/nXI8v

**Comment by maks** [comment]
@RAWJUNGLE sure I'm happy to test it out, but you have set a password on the zip file that is inside the outer zip file.

**Comment by RAWJUNGLE** [comment]
@maks when you download it, you're shown the password 123

**Comment by democloid** [comment]
@RAWJUNGLE You seem to be having this problem trying to read back this SD card on the computer. Do you see the same corruption issues or any errors if you subsequently keep using the SD card in picoTracker?

**Comment by democloid** [comment]
I can't reproduce this. I noticed some samples have spaces in the filename and thought that might be a problem but it's not. Also using OSX and SD card can be read perfectly well after loading samples into project and saving the project.

Any specific samples you're using? Please tell us one specific sample that causes this. Please try to provide a minimal case that causes this, if it happens after loading a single sample, then which one it is? Is it happening to any samples being used?

**Comment by RAWJUNGLE** [comment]
@democloid Pay attention to the posts above, I uploaded the sample pack I use to a file-sharing site. The password for the archive is 123.

**Comment by democloid** [comment]
Of course I read the posts. The sample pack has lots of samples, loaded some random ones and couldn't repro. So asking if it happens with specific samples and to provide the name of one with which it does.

**Comment by RAWJUNGLE** [comment]
@democloid 
1. Format sd to fat32
2. Upload samplelib to sd 
3. Create new project
4. Upload samples to tracks
5. Listen more samples in browser.
6. Play patterns 
7. Save project
8. Off lgpt
9. Insert sd card to computer
10. Upload new samples or folders

**Comment by RAWJUNGLE** [comment]
@democloid ![image](https://github.com/democloid/picoTracker/assets/108550345/655a0273-64e9-46a8-abe9-aa2a42111d4c)

**Comment by RAWJUNGLE** [comment]
@democloid upload to new sd card sandisk 32gb. The problem with the broken file system is gone. But there are still problems with large number of files in the browser and pre-listening.

**Comment by maks** [comment]
@RAWJUNGLE Thanks for continuing to test and update us with your findings.
It sounds like it was a faulty sdcard and the second 32GB  is working for you without filesystem corruption.

In regards to what you see in the photos with the file listing being cut off for directories with many files and then showing just "F" on each line after some limit is reached (on my test dir its 26 files), I can confirm I see exactly the same thing as I mentioned in my previous comment above. Interestingly though, I have a top level directory with 65 sub directories (not files in the top level dir) and all 65 subdirs *are* displayed correctly.

I also can confirm that audio playback does get corrupted when using `listen` button from the sample file browser, but as you saw that is already being tracked in #6. So I think lets keep the focus of this issue now on the bug of listing files in directories which have more than

**Comment by RAWJUNGLE** [comment]
@maks sd 16gb works fine on m8, raspberry pi 4, micro dexed touch, and smartphone. 

I don't know what the problem is.

**Comment by maks** [comment]
@RAWJUNGLE Sorry to keep asking more questions, I'm just trying to better understand what you are seeing...
So that 16gb card can be read/written fine on all those other devices after you save a picotracker project on it but it can't be read on your Mac? 
Or do you mean that after reformatting it, it works fine on all devices until you use it to save a project with the picotracker?

**Comment by RAWJUNGLE** [comment]
@maks Above I described the steps that should cause the problem.
Download the samplelib. And do the steps.
Check it out.

**Comment by democloid** [comment]
@RAWJUNGLE I tried again, using some specific samples that I see in the screenshots and following the steps you provided but still cannot repro. May have to do with the specific SD card for some reason.

**Comment by RAWJUNGLE** [comment]
@democloid There is also a problem with the display of files @maks He confirmed above.

I figured out the file system, just changed the sd card.

**Comment by maks** [comment]
Ok a quick check of adding:
```
#define MAX_FILE_COUNT 24
  int count = 0;  
  FsBaseFile entry;
  while (entry.openNext(&dir, O_READ) && count < MAX_FILE_COUNT) {
```
in `picoTrackerFileSystem.cpp` prevents the `F` chars and no display when scrolling past 25 files I was seeing, so it seems that it is a memory corruption issue due to too large usage mem usage, because going into a dir causes the above function to create a List object of `Path` entries for every file in that directory.

Unfortunately the first thing that comes to mind to fix this: changing to paging through a dirs filelist with the page size being limited to the 15 max lines that can be displayed at a time doesn't look easy to do the way `ImportSampleDialog::setCurrentFolder` is structured at the moment, especially as the code in there also does a alphabetic sort of the file listing :-( 

Another memory saving alternative would be to maybe have a subclass of FilePath that just holds an int index of some kind to each file and then only creates the filename string on demand. 

In any case, this is not a hard fix, but it is a labour intensive one that will need a fair bit of code refactoring unless some one cna think of a clear hack to get around this large ram usage when listing directories of sample files.

**Comment by democloid** [comment]
Storing a sorted list of file indexes would not occupy much memory and sorting the indexes could probably be done by reading the file name from the filesystem for each operation without much impact, but need to test. It does require some laborious refactor, agreed.

Ultimately tho, hard limits need to be set. A linked list is probably not the best solution in this case, so maybe those list objects should be replaced by some other fixed memory size data struct, in order to be able to statically define it and avoid dynamic allocations.

**Comment by RAWJUNGLE** [comment]
@democloid @maks Let's create a chat in Telegram, so that it would be more convenient and faster to communicate directly! Something like a chat of testers.

**Comment by maks** [comment]
awesome thanks for setting up the server @democloid !

In regards to this issue I wanted to propose: 
1. Put in a quick fix straight away, basically I would do I PR with my code above plus add a "fake" File entry in the 26th spot of "[FILE LIMIT REACHED]" and add to existing `docs/MANUAL.md` clearly stating the 25 files per dir limit.
2. Work on larger task of refactoring the sample browser and filesystem adapter to use much less ram for browsing larger sample directories

**Comment by democloid** [comment]
That's a perfectly reasonable short term solution IMO, go for it
